examples/step-86/step-86.cc

```diff
@@ -0,0 +1,896 @@
+              residual[c.index] = y[c.index] - tmp_solution[c.index];
+            }
+          else
+            {
+              // This is a hanging node.
+              residual[c.index] = y[c.index];
+            }
+        }
+    residual.compress(VectorOperation::insert);
+  }
+
+
+  template <int dim>
+  void HeatEquation<dim>::assemble_implicit_jacobian(
+    const double,
+    const PETScWrappers::MPI::Vector &,
+    const PETScWrappers::MPI::Vector &,
+    const double shift)
+  {
+    TimerOutput::Scope t(computing_timer, "assemble_implicit_jacobian");
+    QGauss<dim>        quadrature_formula(fe.degree + 1);
+    FEValues<dim>      fe_values(fe,
+                            quadrature_formula,
+                            update_values | update_gradients |
+                              update_quadrature_points | update_JxW_values);
+
+    const unsigned int dofs_per_cell = fe.n_dofs_per_cell();
+
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+    FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+
+    jacobian_matrix = 0;
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit(cell);
+
+          cell->get_dof_indices(local_dof_indices);
+
+          cell_matrix = 0;
+          for (const unsigned int q : fe_values.quadrature_point_indices())
+            for (const unsigned int i : fe_values.dof_indices())
+              for (const unsigned int j : fe_values.dof_indices())
+                {
+                  cell_matrix(i, j) +=
+                    (shift * fe_values.shape_value(i, q) *
+                       fe_values.shape_value(j, q) +
+                     fe_values.shape_grad(i, q) * fe_values.shape_grad(j, q)) *
+                    fe_values.JxW(q);
+                }
+          constraints.distribute_local_to_global(cell_matrix,
+                                                 local_dof_indices,
+                                                 jacobian_matrix);
+        }
+    jacobian_matrix.compress(VectorOperation::add);
+    // For any constrained degree of freedom, we set the diagonal of the
+    // jacobian to be one. This makes the jacobian matrix invertible, consistent
+    // with what the time stepper expects, and it also makes sure that if we did
+    // not make a mistake in the residual and/or in the jacbian matrix, then
+    // asking the time stepper to check the jacobian with a finite difference
+    // method will produce the correct result. This can be activated at run time
+    // via passing the `-snes_test_jacobian` option on the command line.
+    for (const auto &c : constraints.get_lines())
+      jacobian_matrix.set(c.index, c.index, 1.0);
+    jacobian_matrix.compress(VectorOperation::insert);
+  }
+
+  // Whenever the time stepper solves a linear system, it calls this function to
+  // make sure that hanging nodes and boundary conditions are correctly applied.
```

Comment:
Here  boundary values should be already correct

Reply:
They are correct because we call this function. :)

Reply:
@bangerth This old comment still holds: bc values should be correct at this point (because we solve for the equation y - bc = 0 on them). On the other hand, hanging nodes are incorrect because we don't solve for them, and thus, they need to be updated here.